rlog/2024-04-20-rln-light-verifiers.mdx

```diff
@@ -0,0 +1,173 @@
+
+Following are the requirements and constraints for the solution proposed in this post:
+
+1. Cheap membership insertions.
+2. As few RPC calls as possible to reduce startup time.
+3. Merkle root of the tree is available on-chain.
+4. No centralized services to sequence membership insertions.
+5. Map inserted commitments to the block in which they were inserted.
+
+## Metrics on Sync Time for a tree with 2653 leaves
+
+The following metrics are based on the current implementation of RLN in the Waku gen0 network.
+
+### Test bench
+
+- Hardware: Macbook Air M2, 16GB RAM
+- Network: 120 Megabits/sec
+- Nwaku Commit: [e61e4ff](https://github.com/waku-org/nwaku/tree/e61e4ff90a235657a7dc4248f5be41b6e031e98c)
+- RLN Membership set contract: [0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4](https://sepolia.etherscan.io/address/0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4#code)
+- Deployed block number: 4,230,716
+- RLN Membership set depth: 20
+- Hash function: PoseidonT3 (which is a gas guzzler)
+- Max size of the membership set: 2^20 = 1,048,576 leaves
+
+### Metrics
+
+- Time to sync the whole tree: 4 minutes
+- RPC calls: 702
+- Number of leaves: 2653
+
+One can argue that the time to sync the tree at the current state is not _that_ bad.
+However, the number of RPC calls is a concern,
+which scales linearly with the number of blocks since the contract was deployed.
+This is because the implementation fetches all events from the contract,
+chunking 2000 blocks at a time.
+This is done to avoid hitting the block limit of 10,000 events per call,
+which is a limitation of popular RPC providers.
+
+## Proposed solution
+
+From a theoretical perspective,
+one could construct the merkle tree on-chain,
+in a view call, in-memory.
+However, this is not feasible due to the gas costs associated with it.
+
+To compute the root of a Merkle tree with $2^{20}$ leaves,
+it costs approximately 2 Billion Gas.
+With Infura and Alchemy capping the gas limit to 350M and 550M gas respectively,
+it is not possible to compute the root of the tree in a single call.
+
+Acknowledging that [Polygon Miden](https://polygon.technology/blog/polygon-miden-state-model) and [Penumbra](https://penumbra.zone/blog/tiered-commitment-tree/) both make use of a tiered commitment tree,
+we propose a similar approach for RLN.
+
+A tiered commitment tree is a tree which is sharded into multiple smaller subtrees,
+each of which is a tree in itself.
+This allows scaling in terms of the number of leaves,
+as well as reducing state bloat,
+by just storing the root of a subtree when it is full,
+instead of all its leaves.
+
+Here, the main question arises:
+What is the maximum number of leaves in a subtree, with which the root can be computed in a single call?
+
+To compute the root of a Merkle tree with $2^{10}$ leaves, it costs approximately 217M gas.
+
+This is a feasible number for a single call,
+and hence we propose a tiered commitment tree with a maximum of $2^{10}$ leaves in a subtree,
```

Comment:
remove comma between 'subtree' and 'and'

Reply:
addressed in 66f76f8